social-tw/social-tw-website
# [Test] test: post smart contract

## Summary

Added posting contract unit test and refactor attester functions into another describe function. Installed a dependency to avoid assertion warning in `chai`.

## Linked Issue

#31  

## Details

* added "crypto" to generate `hashUserId`
* added "should revert with invalid proof"
* added some todo comments in `contracts/UnirepApp.sol`
* refactor the structures in `Unirep.test.ts`
* installed `@nomiclabs/hardhat-waffle` in dev dependency to avoid assertion warning in `chai`

## Impacted Areas

* `./contracts/package.json`
* `./contracts/contracts/UnirepApp.sol`
* `./contracts/test/UnirepApp.test.ts`
* `./contracts/type.d.ts`
* `./contracts/tsconfig.json`

## Tests

* "should fail to post with invalid proof"
* "should succeed to post with valid proof"
* "should fail to post with reused proof"

## Possible Impacts

You should install `@nomiclabs/hardhat-waffle` in `packages/contract`.
When integrating with #32, there might be conflicts in `attesters` in `UnirepApp.test.ts` and in the future, posting and logging in should be separated into different files. 

## Visual Materials

-

## Verification Steps

After running `yarn install` and `yarn build`, run `yarn contracts test` in the root dir.
Result should be as follows
```
yarn run v1.22.19
$ yarn workspace @unirep-app/contracts run test
$ hardhat test --network hardhat


  Unirep App
-----------------------------------------------------------------
Epoch tree depth: 17
State tree depth: 17
History tree depth: 17
Number of epoch keys per epoch: 3
Total fields per user: 6
Sum fields per user: 4
Replacement field nonce bits: 48
Replacement field data bits: 205
-----------------------------------------------------------------
Make sure these match what you expect!
-----------------------------------------------------------------
Deploying EpochKeyVerifier
Deploying ProveReputationVerifier
Deploying UserStateTransitionVerifier
Deploying SignupVerifier
Deploying EpochKeyLiteVerifier
Deploying PreventDoubleActionVerifier
Deploying Unirep
-----------------------------------------------------------------
Bytecode size of Unirep: 16184 bytes
Gas cost of deploying Unirep: 3385237
Deployed to: 0x4D137bb44553d55AE6B28B5391c6f537b06C9cc3
-----------------------------------------------------------------
Deploying EpochKeyVerifier
Deploying EpochKeyVerifierHelper
    ✔ deployment (2162ms)
    ✔ init user status & sign up (2160ms)
    user post
      ✔ should fail to post with invalid proof (1347ms)
      ✔ should post with valid proof
      ✔ should fail to post with reused proof
    attesters
Deploying EpochKeyVerifier
Deploying EpochKeyVerifierHelper
      ✔ submit attestations (1733ms)
      ✔ user state transition (16089ms)
Epoch 0 ended
      ✔ data proof (1759ms)


  8 passing (26s)

✨  Done in 27.58s.
```

## Todo

Implement the Smart Contracts tests for posting. The following are the tests we discussed.

- [X]  Happy Path: Post Success
    - [X] Check Event Emit
- [X] Should fail with wrong proof
    - [X] Check revert with the right message
- [X] Should fail with used proof
- [ ] Should write a document

Note: All tests should pass CICD

## Checklist
